Regression test ansi-c/float_constant1 is failing on develop

<!---
Thank you for reporting a problem and suggesting improvements. Please provide the below information to make sure we can effectively deal with the issue reported. For the most precise version information, see the first line of console output or run with --version. Please attach or include example code that allows us to reproduce the problem.
--->

The test `regression/ansi-c/float_constant1` gives me this output:

```
main.c:16:1: error: array size must not be negative, but got -1
 STATIC_ASSERT(__builtin_types_compatible_p(_Float32, __typeof(1.0f32)));
CONVERSION ERROR
EXIT=1
SIGNAL=0
```

This worked before 72957a9128d597e3d2f315df39f5396232f2a53c (which does make changes relating to floating point numbers, but nothing that immediately jumps out as "wrong" to me)

Note that so far I've only seen this happen on machines that have upgraded to Ubuntu 18.04, so maybe this is related to different versions of the default gcc being visible? (`gcc` itself has no issue with `_Float32`, i.e.
```
_Static_assert(__builtin_types_compatible_p(_Float32, __typeof(1.0f32)), "Floating point numbers should work");
``` 
compiles fine with gcc)

--- a/src/ansi-c/literals/convert_float_literal.cpp
+++ b/src/ansi-c/literals/convert_float_literal.cpp
@@ -84,13 +84,21 @@ exprt convert_float_literal(const std::string &src)
   else
     UNREACHABLE;
 
-  const constant_exprt result = a.to_expr();
+  constant_exprt result = a.to_expr();
+  // ieee_floatt::to_expr gives us the representation, but doesn't preserve the
+  // distinction between bitwise-identical types such as _Float32 vs. float,
+  // so ensure we preserve that here:
+  result.type() = type;
 
   if(parsed_float.is_imaginary)
   {
     const complex_typet complex_type(type);
-    return complex_exprt(
-      ieee_floatt::zero(type).to_expr(), result, complex_type);
+
+    constant_exprt zero_real_component = ieee_floatt::zero(type).to_expr();
+    // As above, ensure we preserve the exact type of the literal:
+    zero_real_component.type() = type;
+
+    return complex_exprt(zero_real_component, result, complex_type);
   }
 
   return result;

